src/trace_ng/log.zig

```diff
@@ -0,0 +1,1028 @@
+        pub fn logWithFields(self: *Self, level: Level, message: []const u8, fields: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logWithFields(scope, level, message, fields),
+            }
+        }
+
+        pub fn logfWithFields(self: *Self, level: Level, comptime fmt: []const u8, args: anytype, fields: anytype) void {
+            switch (self.*) {
+                .noop => {},
+                inline else => |impl| impl.logfWithFields(scope, level, fmt, args, fields),
+            }
+        }
+    };
+}
+
+pub const Logger = ScopedLogger(null);
+
+/// An instance of `ScopedLogger` that logs to the standard err.
+const StandardErrLogger = struct {
+    const Self = @This();
+    max_level: Level,
+    exit_sig: std.atomic.Value(bool),
+    allocator: Allocator,
+    log_allocator: Allocator,
+    log_allocator_state: *RecycleFBA(.{}),
+    max_buffer: u64,
+    channel: *Channel(logfmt.LogMsg),
+    handle: ?std.Thread,
+
+    pub fn init(config: Config) !*Self {
+        const recycle_fba = try config.allocator.create(RecycleFBA(.{}));
+        const max_buffer = config.max_buffer orelse return error.MaxBufferNotSet;
+        recycle_fba.* = try RecycleFBA(.{}).init(config.allocator, max_buffer);
+        const self = try config.allocator.create(Self);
+        self.* = .{
+            .allocator = config.allocator,
+            .log_allocator = recycle_fba.allocator(),
+            .log_allocator_state = recycle_fba,
+            .max_buffer = max_buffer,
+            .exit_sig = AtomicBool.init(false),
+            .max_level = config.max_level,
+            .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+            .handle = try std.Thread.spawn(.{}, run, .{self}),
+        };
+        return self;
+    }
+
+    pub fn deinit(self: *Self) void {
+        self.channel.close();
+        if (self.handle) |*handle| {
+            std.time.sleep(std.time.ns_per_ms * 5);
+            self.exit_sig.store(true, .seq_cst);
+            handle.join();
+        }
+        self.channel.deinit();
+        self.log_allocator_state.deinit();
+        self.allocator.destroy(self.log_allocator_state);
+        self.allocator.destroy(self);
+    }
+
+    pub fn unscoped(self: Self) Logger {
+        return .{ .standard = self };
+    }
+
+    pub fn withScope(self: Self, comptime new_scope: anytype) ScopedLogger(new_scope) {
+        return .{ .standard = self };
+    }
+
+    pub fn run(self: *Self) void {
+        while (!self.exit_sig.load(.seq_cst)) {
+            std.time.sleep(std.time.ns_per_ms * 5);
+            const messages = self.channel.drain() orelse {
+                // channel is closed
+                return;
+            };
+            defer self.channel.allocator.free(messages);
+            for (messages) |message| {
+                const writer = std.io.getStdErr().writer();
+                std.debug.lockStdErr();
```

Comment:
I think the base case is to acquire the lock for each message written. This is what the std lib log does (alongside using buffered writer), and also debug.print. And given that we are offloading the writing out of the log message off to another thread, hopefully the computation needed to acquire/release the std err does not easily affect other working threads.

Also giving that the stderr is a shared resource, i'll err on the side of not holding on to it too much. Even though the likelihood of that happening is low in this case, if there is a bug in the logger, where the loop runs for more time than needed, there is the possibility of hogging unto the shared resource for more than needed.

So maybe acquire the lock for each message (as it is currently) but if in practice, this proofs to have a noticeable performance impact, then it can be revisited?